### PR TITLE
Catch and ignore AttributeError for module.__file__.

### PR DESCRIPTION
**Pull request by mauritsvanrees**
Fixes https://github.com/zopefoundation/z3c.autoinclude/issues/6.

### CODE REVIEW

**Comment by mauritsvanrees** [comment]
I tried triggering this with extra test code, but failed at it. To not pollute the current PR branch with non-useful tests I have created a new branch for that. See the commits on the[maurits-attributeerror-module-file-issue-6-trying-tests](https://github.com/zopefoundation/z3c.autoinclude/commits/maurits-attributeerror-module-file-issue-6-trying-tests) branch.

**Comment by mauritsvanrees** [comment]
I *can* trigger the error locally now. I have Python 2.7 installed from the [Anaconda distribution](https://www.anaconda.com/distribution/) and this apparently has the `backports` package, which triggers it.
When I create a virtualenv in the `z3c.autoinclude` checkout and run the tests just like `.travis.yml` sets them up, all is well. But when I do `python2.7 bootstrap.py && bin/buildout && bin/test -v1`, the problem is triggered (and the PR branch fixes it).
For some reason, the tests take *really* long then: almost half an hour!

A pdb in `dependency.py includableInfo` (a few lines earlier than I meant, but okay) shows something interesting:

```
/anaconda2/lib/python2.7/site-packages/astropy/config/configuration.py:541: ConfigurationMissingWarning: Configuration defaults will be used due to OSError:2 on None
  warn(ConfigurationMissingWarning(msg))
z3c.autoinclude: 'backports' has no __file__ attribute  <---- Hah! My fix works, and we only get a warning here.
> /Users/maurits/community/plone-coredev/5.2/src/z3c.autoinclude/src/z3c/autoinclude/dependency.py(33)includableInfo()
-> logger.warning(
(Pdb) l
 28  	            for dotted_name in dist_manager.dottedNames():
 29  	                try:
 30  	                    module = resolve(dotted_name)
 31  	                except ImportError as exc:
 32  	                    import pdb; pdb.set_trace()
 33  ->	                    logger.warning(
 34  	                        "resolve(%r) raised import error: %s" % (dotted_name, exc))
 35  	                    continue
 36  	                module_file = getattr(module, '__file__', None)
 37  	                if module_file is None:
 38  	                    logger.warning(
(Pdb) a
self = <z3c.autoinclude.dependency.DependencyFinder object at 0x10b4cec90>
zcml_to_look_for = ['configure.zcml', 'meta.zcml']
(Pdb) self.context
BCPackage 0.1 (/private/var/folders/g1/3lr2tj2d1vn80qb686lcbrbm0000gp/T/tmpR0M_ZD.z3c.autoinclude.test-installs/BCPackage-0.1-py2.7.egg)
(Pdb) self.context.requires()
[Requirement.parse('setuptools'), Requirement.parse('TestDirective'), Requirement.parse('SiblingPackage')]
(Pdb) dotted_name
'bkcharts'
(Pdb) req
Requirement.parse('setuptools')
(Pdb) dist_manager
<z3c.autoinclude.utils.DistributionManager object at 0x10be13350>
(Pdb) dist_manager.dottedNames()
['Crypto', 'Cython', 'IPython', 'OpenSSL', 'PIL', 'PyQt5', '_pytest', 'alabaster', 'anaconda_navigator', 'anaconda_project', 'appnope', 'asn1crypto', 'astroid', 'astropy', 'attr', 'babel', 'backports', 'binstar_client', 'bitarray', 'bkcharts', 'blaze', 'bleach', 'bokeh', 'boto', 'bottleneck', 'bs4', 'certifi', 'cffi', 'chardet', 'click', 'cloudpickle', 'clyent', 'colorama', 'concurrent', 'conda', 'conda_build', 'conda_env', 'conda_verify', 'cryptography', 'curl', 'cytoolz', 'dask', 'datashape', 'dateutil', 'distributed', 'doc8', 'docutils', 'enum', 'et_xmlfile', 'fastcache', 'flask', 'flask_cors', 'funcsigs', 'functools32', 'gevent', 'glob2', 'h5py', 'html5lib', 'idna', 'imageio', 'ipykernel', 'ipython_genutils', 'ipywidgets', 'isort', 'jedi', 'jinja2', 'jsonschema', 'jupyter_client', 'jupyter_console', 'jupyter_core', 'jupyterlab', 'jupyterlab_launcher', 'lazy_object_proxy', 'llvmlite', 'locket', 'lxml', 'markupsafe', 'matplotlib', 'mkl', 'more_itertools', 'mpmath', 'msgpack', 'multipledispatch', 'navigator_updater', 'nbconvert', 'nbformat', 'networkx', 'nltk', 'nose', 'notebook', 'numba', 'numexpr', 'numpy', 'numpydoc', 'odo', 'olefile', 'openpyxl', 'packaging', 'pandas', 'parso', 'partd', 'pathlib2', 'patsy', 'pbr', 'pexpect', 'pip', 'pkg_resources', 'pkginfo', 'pluggy', 'ply', 'prompt_toolkit', 'psutil', 'ptyprocess', 'py', 'pycparser', 'pyflakes', 'pygments', 'pylint', 'pytz', 'pywt', 'pyximport', 'qtawesome', 'qtconsole', 'qtpy', 'requests', 'restructuredtext_lint', 'rope', 'ruamel_yaml', 'scipy', 'scripts', 'seaborn', 'send2trash', 'setuptools', 'skimage', 'sklearn', 'snowballstemmer', 'sortedcollections', 'sortedcontainers', 'sphinx', 'sphinxcontrib', 'spyder', 'spyder_breakpoints', 'spyder_io_dcm', 'spyder_io_hdf5', 'spyder_profiler', 'spyder_pylint', 'sqlalchemy', 'statsmodels', 'stevedore', 'sympy', 'tables', 'tblib', 'terminado', 'testpath', 'tlz', 'toolz', 'tornado', 'traitlets', 'unicodecsv', 'urllib3', 'wcwidth', 'webencodings', 'werkzeug', 'wheel', 'widgetsnbextension', 'wrapt', 'xlrd', 'xlsxwriter', 'xlwings', 'xlwt', 'yaml', 'zict', 'zmq']
(Pdb) len(dist_manager.dottedNames())
173
```

So the test `BCPackage`, which only has three packages in its `install_requires`, gets a list of 173 packages in `dist_manager.dottedNames`. (This is the dist manager around the setuptools package). In don't quite know what that is, but I would have expected only three in there, not 170 more. This explains why the tests take so ridiculously long this way.
I suppose I should use a cleaner and simpler Python 2.7. :-)
So it looks like our `utils.DistributionManager` class finds way too much. That sounds like a different bug.

Result is that several other warnings or errors are printed. Just a few of them:

```
z3c.autoinclude: resolve('bkcharts') raised import error: cannot import name DEFAULT_PALETTE
/anaconda2/lib/python2.7/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
/anaconda2/lib/python2.7/site-packages/multipledispatch/dispatcher.py:35: DeprecationWarning: halt_ordering is deprecated, you can safely remove this call.
  DeprecationWarning,
/anaconda2/lib/python2.7/site-packages/scipy/spatial/__init__.py:96: ImportWarning: Not importing directory '/anaconda2/lib/python2.7/site-packages/scipy/spatial/qhull': missing __init__.py
  from .qhull import *
/anaconda2/lib/python2.7/site-packages/odo/backends/pandas.py:102: FutureWarning: pandas.tslib is deprecated and will be removed in a future version.
You can access NaTType as type(pandas.NaT)
  @convert.register((pd.Timestamp, pd.Timedelta), (pd.tslib.NaTType, type(None)))
...
z3c.autoinclude: 'sphinxcontrib' has no __file__ attribute
```

Anyway, my conclusion: this PR fixes problems that can apparently easily occur when using a non-clean, non-virtualenv Python.

**Comment by mauritsvanrees** [comment]
And just for the record: no this can't be triggered with a virtualenv by installing `backports`, because it does not contain any code.  It is only registered on PyPI to prevent someone accidentally or maliciously claiming the `backports` namespace.

```
$ bin/pip install backports
Collecting backports
  Could not find a version that satisfies the requirement backports (from versions: )
No matching distribution found for backports
```

And installing one of the real backports namespace packages in a virtualenv (`pip install backports.shutil-which`) works fine, but does not trigger the problem this PR solves.

As indicated, I tried recreating a [similar situation](https://github.com/zopefoundation/z3c.autoinclude/blob/maurits-attributeerror-module-file-issue-6-trying-tests/src/z3c/autoinclude/tests/backportsclone/setup.py) in the tests, on my second branch, but this failed.

**Comment by mgedmin** [comment]
So it's namespace packages that trigger the bug?

Would a synthetic test where you stuff some object with no `__file__` attribute into sys.modules[] suffice to trigger this code?  

(Please don't feel pressured to waste too much time on this!)

**Comment by mauritsvanrees** [comment]
Hah, I have found a way with some mild hacking.

When running the tests from `dependency.txt`, you know get a few warnings, indicating that my fix is active and working, and a similar check is working as well:

```
/Users/maurits/shared-eggs/cp27m/zope.dottedname-4.3-py2.7.egg/zope/dottedname/resolve.py:45: ImportWarning: Not importing directory '/Users/maurits/community/plone-coredev/5.2/src/z3c.autoinclude/src/z3c/autoinclude/tests/empty': missing __init__.py
  __import__(used)
z3c.autoinclude: resolve('z3c.autoinclude.tests.empty') raised import error: No module named empty
z3c.autoinclude: resolve('nosuchpackage') raised import error: No module named nosuchpackage
z3c.autoinclude: resolve('z3c.autoinclude.tests.empty') raised import error: No module named empty
z3c.autoinclude: resolve('nosuchpackage') raised import error: No module named nosuchpackage
z3c.autoinclude: resolve('z3c.autoinclude.tests.empty') raised import error: No module named empty
z3c.autoinclude: resolve('nosuchpackage') raised import error: No module named nosuchpackage
```

(I forgot to commit the `tests/empty/.gitkeep` file, so Travis might initially fail, but I force pushed it just now, so that should fix it again.)

**Comment by mauritsvanrees** [comment]
Ah, no: my hacking only tests the other already existing check for an `ImportError`.
Let me see if I can rework my test.

**Comment by mauritsvanrees** [comment]
Got it.
Without my fix, the tests now say this:

```
Failure in test /Users/maurits/community/plone-coredev/5.2/src/z3c.autoinclude/src/z3c/autoinclude/tests/../dependency.txt
Failed doctest test for dependency.txt
  File "/Users/maurits/community/plone-coredev/5.2/src/z3c.autoinclude/src/z3c/autoinclude/tests/../dependency.txt", line 0

----------------------------------------------------------------------
File "/Users/maurits/community/plone-coredev/5.2/src/z3c.autoinclude/src/z3c/autoinclude/tests/../dependency.txt", line 171, in dependency.txt
Failed example:
    pprint(b_include_finder.includableInfo(['configure.zcml',
                                            'meta.zcml']))
Exception raised:
    Traceback (most recent call last):
      File "/usr/local/py27/parts/opt/lib/python2.7/doctest.py", line 1315, in __run
        compileflags, 1) in test.globs
      File "<doctest dependency.txt[40]>", line 2, in <module>
        'meta.zcml']))
      File "/Users/maurits/community/plone-coredev/5.2/src/z3c.autoinclude/src/z3c/autoinclude/dependency.py", line 33, in includableInfo
        os.path.dirname(module.__file__), candidate)
    AttributeError: 'module' object has no attribute '__file__'
```

With my fix there are only some warnings:

```
z3c.autoinclude: 'z3c.autoinclude.tests.nomodulefile' has no __file__ attribute
z3c.autoinclude: resolve('nosuchpackage') raised import error: No module named nosuchpackage
z3c.autoinclude: 'z3c.autoinclude.tests.nomodulefile' has no __file__ attribute
z3c.autoinclude: resolve('nosuchpackage') raised import error: No module named nosuchpackage
z3c.autoinclude: 'z3c.autoinclude.tests.nomodulefile' has no __file__ attribute
z3c.autoinclude: resolve('nosuchpackage') raised import error: No module named nosuchpackage
```

**Comment by mgedmin** [comment]
Nice!

File: src/z3c/autoinclude/tests/nomodulefile/__init__.py [hunk 1]
```diff
@@ -0,0 +1,6 @@
+# -*- coding: utf-8 -*-
+# This directory is used in a test in dependency.txt.
+# We need a package with a module that has no __file__ attribute.
+# This can happen in corner cases, for example with 'backports' on Python 2.7.
+# See https://github.com/zopefoundation/z3c.autoinclude/issues/6.
+del __file__
```
Comment (mgedmin) [inline src/z3c/autoinclude/tests/nomodulefile/__init__.py]: Haha awesome!

**Comment by mauritsvanrees** [comment]
The PR is approved, so I merge it.
